File: src/Isr.cc
// -*- lsst-c++ -*-

/*
 * LSST Data Management System
 * Copyright 2008, 2009, 2010 LSST Corporation.
 *
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the LSST License Statement and
 * the GNU General Public License along with this program.  If not,
 * see <http://www.lsstcorp.org/LegalNotices/>.
 */

#include <cmath>

#include "lsst/geom.h"
#include "lsst/afw/cameraGeom.h"
#include "lsst/afw/math.h"
#include "lsst/afw/math/Statistics.h"
#include "lsst/ip/isr/isr.h"

namespace lsst { namespace ip { namespace isr {

template <typename PixelT>
size_t maskNans(afw::image::MaskedImage<PixelT> const& mi, afw::image::MaskPixel maskVal,
                afw::image::MaskPixel allow)
{
    typedef typename afw::image::MaskedImage<PixelT>::x_iterator x_iterator;
    size_t nPix = 0;
    for (int y = 0; y != mi.getHeight(); ++y) {
        for (x_iterator ptr = mi.row_begin(y), end = mi.row_end(y); ptr != end; ++ptr) {
            if (!(ptr.mask() & allow) && (!std::isfinite(ptr.image()) ||
                                          !std::isfinite(ptr.variance()))) {
                nPix += 1;
                ptr.mask() |= maskVal;
            }
        }
    }
    return nPix;
}

template<typename ImagePixelT>
std::vector<double> fitOverscanImage(
    afw::image::MaskedImage<ImagePixelT> const& overscan,
    std::vector<std::string> badPixelMask,
    bool isTransposed
) {
    typedef afw::image::MaskedImage<ImagePixelT> MaskedImage;

    /**
    This is transposed here to match the existing numpy-array ordering.
    This effectively transposes the image for us.
    **/
    const int height = overscan.getHeight();
    const int width  = overscan.getWidth();

    int length = height;
    if (isTransposed) {
        length = width;
    }

    std::vector<double> values(length);

    afw::math::StatisticsControl statControl;
    statControl.setAndMask(overscan.getMask()->getPlaneBitMask(badPixelMask));

    const int x0 = overscan.getX0();
    const int y0 = overscan.getY0();
    auto origin = geom::Point2I(x0, y0);
    geom::Extent2I shifter;
    geom::Extent2I extents;
    if (isTransposed) {
        shifter = geom::Extent2I(1, 0);
        extents = geom::Extent2I(1, height);
    } else {
        shifter = geom::Extent2I(0, 1);
        extents = geom::Extent2I(width, 1);
    }

    for (int x = 0; x < length; ++x) {
        MaskedImage mi = MaskedImage(overscan, geom::Box2I(origin, extents));
        values[x] = afw::math::makeStatistics(mi, afw::math::MEDIAN, statControl).getValue();
        origin.shift(shifter);
    }
    return values;
}

template<typename ImagePixelT>
std::vector<double> fitOverscanImageMean(
    afw::image::MaskedImage<ImagePixelT> const& overscan,
    std::vector<std::string> badPixelMask,
    bool isTransposed
) {
    typedef afw::image::MaskedImage<ImagePixelT> MaskedImage;

    /**
    This is transposed here to match the existing numpy-array ordering.
    This effectively transposes the image for us.
    **/
    const int height = overscan.getHeight();
    const int width  = overscan.getWidth();

    int length = height;
    if (isTransposed) {
        length = width;
    }

    std::vector<double> values(length);

    afw::math::StatisticsControl statControl;
    statControl.setAndMask(overscan.getMask()->getPlaneBitMask(badPixelMask));

    const int x0 = overscan.getX0();
    const int y0 = overscan.getY0();
    auto origin = geom::Point2I(x0, y0);
    geom::Extent2I shifter;
    geom::Extent2I extents;
    if (isTransposed) {
        shifter = geom::Extent2I(1, 0);
        extents = geom::Extent2I(1, height);
    } else {
        shifter = geom::Extent2I(0, 1);
        extents = geom::Extent2I(width, 1);
    }

    for (int x = 0; x < length; ++x) {
        MaskedImage mi = MaskedImage(overscan, geom::Box2I(origin, extents));
        values[x] = afw::math::makeStatistics(mi, afw::math::MEAN, statControl).getValue();
        origin.shift(shifter);
    }
    return values;
}

std::string between(std::string &s, char ldelim, char rdelim) {
    std::string::iterator b(s.begin());
    std::string::iterator e(s.end());
    std::string::iterator lp;
    std::string::iterator rp;

    std::string result;

    if((lp = std::find(b, e, ldelim)) != e)
        if((rp = std::find(++lp, e, rdelim)) != e)
            result = std::string(lp, rp);

    return result;
}

template<typename ImagePixelT>
lsst::afw::image::MaskedImage<ImagePixelT> computeCrosstalkSubtrahend(
    lsst::afw::image::Exposure<ImagePixelT> const& exp, ///< Input exposure
    ndarray::Array<double, 2> const& coeffs, ///< Crosstalk coefficients
    ndarray::Array<double, 2> const& coeffsSqr, ///< Nonlinear Crosstalk coefficients
    bool applyMask ///< Transfer mask as well?  RENAME copyCrosstalkMask
) {
    typedef std::shared_ptr<afw::image::Image<ImagePixelT>> ImagePtr;
    typedef afw::image::MaskedImage<ImagePixelT> MaskedImage;

    MaskedImage subtrahend = MaskedImage(exp.getBBox());
    subtrahend.getImage() = 0;
    subtrahend.getMask() = 0;
    subtrahend.getVariance() = 0;

    std::map<lsst::afw::cameraGeom::ReadoutCorner, bool> X_FLIP;
    std::map<lsst::afw::cameraGeom::ReadoutCorner, bool> Y_FLIP;

    X_FLIP[lsst::afw::cameraGeom::ReadoutCorner::LL] = false;
    X_FLIP[lsst::afw::cameraGeom::ReadoutCorner::LR] = true;
    X_FLIP[lsst::afw::cameraGeom::ReadoutCorner::UL] = false;
    X_FLIP[lsst::afw::cameraGeom::ReadoutCorner::UR] = true;

    Y_FLIP[lsst::afw::cameraGeom::ReadoutCorner::LL] = false;
    Y_FLIP[lsst::afw::cameraGeom::ReadoutCorner::LR] = false;
    Y_FLIP[lsst::afw::cameraGeom::ReadoutCorner::UL] = true;
    Y_FLIP[lsst::afw::cameraGeom::ReadoutCorner::UR] = true;

    auto isTrimmed = (exp.getDetector()->getBBox() == exp.getBBox());

    auto amplifiers = exp.getDetector()->getAmplifiers();
    auto nAmp = amplifiers.size();

    for (size_t sourceIndex = 0; sourceIndex < nAmp; ++sourceIndex) {
        for (size_t targetIndex = 0; targetIndex < nAmp; ++targetIndex) {
            auto coeff = coeffs[sourceIndex][targetIndex];
            auto coeffSqr = coeffsSqr[sourceIndex][targetIndex];

            // This should be 0.0 if it needs to be skipped (bad amp, etc).
            if (coeff == 0.0) {
                continue;
            }

            auto sourceAmp = amplifiers[sourceIndex];
            auto targetAmp = amplifiers[targetIndex];

            lsst::geom::Box2I sourceBBox, targetBBox;

            if (isTrimmed) {
                sourceBBox = sourceAmp->getBBox();
                targetBBox = targetAmp->getBBox();
            } else {
                sourceBBox = sourceAmp->getRawBBox();
                targetBBox = targetAmp->getRawBBox();
            }

            auto sourceImageIn = exp[sourceBBox].getMaskedImage().getImage();
            auto targetImage = subtrahend[targetBBox].getImage();
            // We may not need this at all.
            auto sourceMask = exp[sourceBBox].getMaskedImage().getMask();

            auto sourceAmpCorner = sourceAmp->getReadoutCorner();
            auto targetAmpCorner = targetAmp->getReadoutCorner();

            bool xFlip = X_FLIP[targetAmpCorner] ^ X_FLIP[sourceAmpCorner];
            bool yFlip = Y_FLIP[targetAmpCorner] ^ Y_FLIP[sourceAmpCorner];

            // This makes a copy of the amplifier data.
            auto sourceImage = lsst::afw::math::flipImage(*sourceImageIn, xFlip, yFlip);

            // The original takes coeffs[ss, tt].
            // tt -> tAmp -> "amp" which is the one that is extracted.
            // ss -> sAmp -> "targetAmp" which is the one to match (subtrahend).
            // But it doesn't matter exactly, we xor them anyway.
            // But we are flipping the the tAmp which is *not* the subtrahend.  Okay.

            // Here we would subtract the background if we have it.  TBD

            targetImage->scaledPlus(coeff, *sourceImage);
            if (coeffSqr != 0.0) {
                // This is okay because we have already made a copy.
                sourceImage->scaledMultiplies(1.0, *sourceImage);
                targetImage->scaledPlus(coeffSqr, *sourceImage);
            }

        }
    }

    return subtrahend;
}


// Explicit instantiations

template
std::vector<double> fitOverscanImage<int>(
    afw::image::MaskedImage<int> const&, std::vector<std::string> badPixelMask, bool isTransposed);
template
std::vector<double> fitOverscanImage<float>(
    afw::image::MaskedImage<float> const&, std::vector<std::string> badPixelMask, bool isTransposed);
template
std::vector<double> fitOverscanImage<double>(
    afw::image::MaskedImage<double> const&, std::vector<std::string> badPixelMask, bool isTransposed);

template
std::vector<double> fitOverscanImageMean<int>(
    afw::image::MaskedImage<int> const&, std::vector<std::string> badPixelMask, bool isTransposed);
template
std::vector<double> fitOverscanImageMean<float>(
    afw::image::MaskedImage<float> const&, std::vector<std::string> badPixelMask, bool isTransposed);
template
std::vector<double> fitOverscanImageMean<double>(
    afw::image::MaskedImage<double> const&, std::vector<std::string> badPixelMask, bool isTransposed);

template class CountMaskedPixels<float>;
template class CountMaskedPixels<double>;
template class CountMaskedPixels<int>;

// Function to mask nans in a masked image
template size_t maskNans<float>(afw::image::MaskedImage<float> const&, afw::image::MaskPixel,
                                afw::image::MaskPixel);
template size_t maskNans<double>(afw::image::MaskedImage<double> const&, afw::image::MaskPixel,
                                 afw::image::MaskPixel);
template size_t maskNans<int>(afw::image::MaskedImage<int> const&, afw::image::MaskPixel,
                              afw::image::MaskPixel);

template
afw::image::MaskedImage<float> computeCrosstalkSubtrahend<float>(
    lsst::afw::image::Exposure<float> const&,
    ndarray::Array<double, 2> const&,
    ndarray::Array<double, 2> const&,
    bool applyMask=false);

template
afw::image::MaskedImage<double> computeCrosstalkSubtrahend<double>(
    lsst::afw::image::Exposure<double> const&,
    ndarray::Array<double, 2> const&,
    ndarray::Array<double, 2> const&,
    bool applyMask=false);

// This is to make pybind11 and the wrapper happy
template
afw::image::MaskedImage<int> computeCrosstalkSubtrahend<int>(
    lsst::afw::image::Exposure<int> const&,
    ndarray::Array<double, 2> const&,
    ndarray::Array<double, 2> const&,
    bool applyMask=false);

}}} // namespace lsst::ip::isr

File: python/lsst/ip/isr/Isr.py
#
# LSST Data Management System
# Copyright 2008, 2009, 2010 LSST Corporation.
#
# This product includes software developed by the
# LSST Project (http://www.lsst.org/).
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the LSST License Statement and
# the GNU General Public License along with this program.  If not,
# see <http://www.lsstcorp.org/LegalNotices/>.
#

import math, re
import lsst.afw.image as afwImage

class Isr(object):
    def createPsf(fwhm):
        """Make a PSF"""
        ksize = 4*int(fwhm) + 1
        return afwDetection.createPsf('DoubleGaussian', ksize, ksize, fwhm/(2*math.sqrt(2*math.log(2))))

    def calcEffectiveGain(exposure):
        mi = exposure.getMaskedImage()
        im = afwImage.ImageF(mi.getImage(), True)
        var = mi.getVariance()
        im /= var
        medgain = afwMath.makeStatistics(im, afwMath.MEDIAN).getValue()
        meangain = afwMath.makeStatistics(im, afwMath.MEANCLIP).getValue()
        return medgain, meangain

    def convertImageForIsr(exposure):
        if not isinstance(exposure, afwImage.ExposureU):
            raise Exception("ipIsr.convertImageForIsr: Expecting Uint16 image. Got\
                %s."%(exposure.__repr__()))

        newexposure = exposure.convertF()
        amp = cameraGeom.cast_Amp(exposure.getDetector())
        mi = newexposure.getMaskedImage()
        var = afwImage.ImageF(mi.getBBox(afwImage.PARENT))
        mask = afwImage.MaskU(mi.getBBox(afwImage.PARENT))
        mask.set(0)
        newexposure.setMaskedImage(afwImage.MaskedImageF(mi.getImage(), mask, var))
        return newexposure

    def calculateSdqaCcdRatings(exposure):
        metrics = {}
        metrics['nSaturatePix'] = 0
        metrics['nBadCalibPix'] = 0
        metrics['imageClipMean4Sig3Pass'] = None
        metrics['imageSigma'] = None
        metrics['imageMedian'] = None
        metrics['imageMin'] = None
        metrics['imageMax'] = None
        metadata = exposure.getMetadata()
        mi = exposure.getMaskedImage()
        mask = mi.getMask()
        badbitmask = mask.getPlaneBitMask('BAD')
        satbitmask = mask.getPlaneBitMask('SAT')
        intrpbitmask = mask.getPlaneBitMask('INTRP')
        sctrl = afwMath.StatisticsControl()
        sctrl.setNumIter(3)
        sctrl.setNumSigmaClip(4)
        sctrl.setAndMask(satbitmask | badbitmask | intrpbitmask)
        satmask = afwImage.MaskU(mask, True)
        badmask = afwImage.MaskU(mask, True)
        satmask &= satbitmask
        badmask &= badbitmask
        satmaskim = afwImage.ImageU(satmask.getBBox(afwImage.PARENT))
        satmaskim <<= satmask
        badmaskim = afwImage.ImageU(badmask.getBBox(afwImage.PARENT))
        badmaskim <<= badmask
        thresh = afwDetection.Threshold(0.5)
        fs = afwDetection.makeFootprintSet(satmaskim, thresh)
        for f in fs.getFootprints():
            metrics['nSaturatePix'] += f.getNpix()
        fs = afwDetection.makeFootprintSet(badmaskim, thresh)
        for f in fs.getFootprints():
            metrics['nBadCalibPix'] += f.getNpix()
        stats = afwMath.makeStatistics(mi, afwMath.MEANCLIP | \
            afwMath.STDEVCLIP | afwMath.MEDIAN | afwMath.MIN |\
            afwMath.MAX, sctrl)
        metrics['imageClipMean4Sig3Pass'] = stats.getValue(afwMath.MEANCLIP)
        metrics['imageSigma'] = stats.getValue(afwMath.STDEVCLIP)
        metrics['imageMedian'] = stats.getValue(afwMath.MEDIAN)
        metrics['imageMin'] = stats.getValue(afwMath.MIN)
        metrics['imageMax'] = stats.getValue(afwMath.MAX)
        for k in metrics.keys():
            metadata.set(k, metrics[k])

    def calculateSdqaAmpRatings(exposure, biasBBox, dataBBox):
        metrics = {}
        metrics['nSaturatePix'] = 0
        metrics['overscanMean'] = None
        metrics['overscanStdDev'] = None
        metrics['overscanMedian'] = None
        metrics['overscanMin'] = None
        metrics['overscanMax'] = None
        mi = exposure.getMaskedImage()
        metadata = exposure.getMetadata()
        trimmi = afwImage.MaskedImageF(mi, dataBBox, False)
        biasmi = afwImage.MaskedImageF(mi, biasBBox, False)
        mask = mi.getMask()
        satbitmask = mask.getPlaneBitMask('SAT')
        sctrl = afwMath.StatisticsControl()
        sctrl.setAndMask(satbitmask)
        satmask = trimmi.getMask()
        satmask &= satbitmask
        satmaskim = afwImage.ImageU(satmask.getBBox(afwImage.PARENT))
        satmaskim <<= satmask
        thresh = afwDetection.Threshold(0.5)
        fs = afwDetection.makeFootprintSet(satmaskim, thresh)
        for f in fs.getFootprints():
            metrics['nSaturatePix'] += f.getNpix()
        stats = afwMath.makeStatistics(biasmi, afwMath.MEAN | \
            afwMath.STDEV | afwMath.MEDIAN | afwMath.MIN |\
            afwMath.MAX,sctrl)
        metrics['overscanMean'] = stats.getValue(afwMath.MEAN)
        metrics['overscanStdDev'] = stats.getValue(afwMath.STDEV)
        metrics['overscanMedian'] = stats.getValue(afwMath.MEDIAN)
        metrics['overscanMin'] = stats.getValue(afwMath.MIN)
        metrics['overscanMax'] = stats.getValue(afwMath.MAX)
        for k in metrics.keys():
            metadata.set(k, metrics[k])




    def maskFromDefects(dimensions, fpList):
        # the output LSST Mask image
        minpt = afwGeom.Point2I(0,0)
        maxpt = afwGeom.Point2I(dimensions[0], dimensions[1])
        mask = afwImage.MaskU(afwGeom.Box2I(minpt, maxpt), afwImage.LOCAL)
        mask.set(0)
        bitmask = mask.getPlaneBitMask('BAD')

        # set the bits
        afwDetection.setMaskFromFootprintList(mask, fpList, bitmask)

        return mask


    def defectsFromBoolImage(fitsfile, invert=False):
        # input bad pixel image
        # This assumes an image with ones and zeros
        image = afwImage.ImageF(fitsfile)
        if invert:
            image *= -1
            thresh = afwDetection.Threshold(-0.5)
        else:
            thresh = afwDetection.Threshold(0.5)

        # turn into masked image for detection
        mi = afwImage.MaskedImageF(image)

        # find bad regions
        ds = afwDetection.FootprintSetF(mi, thresh)
        fpList = ds.getFootprints()

        return fpList

    def interpolateDefectList(exposure, defectList, fwhm, fallbackValue=None):
        mi = exposure.getMaskedImage()
        psf = createPsf(fwhm)
        if fallbackValue is None:
            fallbackValue = afwMath.makeStatistics(mi.getImage(), afwMath.MEANCLIP).getValue()
        algorithms.interpolateOverDefects(mi, psf, defectList, fallbackValue)

    def maskBadPixelsDef(exposure, defectList, fwhm=None,
                 interpolate=True,
                 maskName='BAD'):

        # mask bad pixels
        mi = exposure.getMaskedImage()
        mask = mi.getMask()
        bitmask = mask.getPlaneBitMask(maskName)
        for defect in defectList:
            bbox = defect.getBBox()
            afwDetection.setMaskFromFootprint(mask, afwDetection.Footprint(bbox), bitmask)

        if interpolate:
            # and interpolate over them
            assert fwhm and fwhm > 0, "FWHM not provided for interpolation"
            interpolateDefectList(exposure, defectList, fwhm)

    def linearization(exposure, lookupTable):

        # common input test
        metadata   = exposure.getMetadata()
        gain = metadata.get('gain')
        mi   = exposure.getMaskedImage()
        lookupTable.apply(mi, gain)

    def saturationDetection(exposure, saturation, doMask = True, maskName = 'SAT'):

        mi = exposure.getMaskedImage()
        if display:
            ds9.mtv(mi, frame=0)


        # find saturated regions
        thresh = afwDetection.Threshold(saturation)
        ds = afwDetection.makeFootprintSet(mi, thresh)
        fpList = ds.getFootprints()
        # we will turn them into defects for interpolating
        defectList = algorithms.DefectListT()

        # grow them
        bboxes = []
        for fp in fpList:
            if doMask:
                mask = mi.getMask()
                bitmask = mask.getPlaneBitMask(maskName)
                afwDetection.setMaskFromFootprint(mask, fp, bitmask)
            bboxes.append(afwDetection.footprintToBBoxList(fp))
        return bboxes

    def saturationInterpolation(exposure, fwhm, growFootprints = 1, maskName = 'SAT'):
        mi = exposure.getMaskedImage()
        mask = mi.getMask()
        satmask = afwImage.MaskU(mask, True)
        satmask &= mask.getPlaneBitMask(maskName)
        thresh = afwDetection.Threshold(0.5)
        maskimg = afwImage.ImageU(satmask.getBBox(afwImage.PARENT))
        maskimg <<= satmask
        ds = afwDetection.makeFootprintSet(maskimg, thresh)
        fpList = ds.getFootprints()
        satDefectList = algorithms.DefectListT()
        for fp in fpList:
            if growFootprints > 0:
                # if "True", growing requires a convolution
                # if "False", its faster
                fpGrow = afwDetection.growFootprint(fp, growFootprints, False)
            else:
                fpGrow = fp
            for bbox in afwDetection.footprintToBBoxList(fpGrow):
                defect = algorithms.Defect(bbox)
                satDefectList.push_back(defect)
        if 'INTRP' not in mask.getMaskPlaneDict().keys():
            mask.addMaskPlane('INTRP')
        psf = createPsf(fwhm)

        algorithms.interpolateOverDefects(mi, psf, satDefectList)


    def saturationCorrection(exposure, saturation, fwhm, growFootprints=1,
                 interpolate = True,
                 maskName    = 'SAT'):
        #This will be slower than necessary.  Should write another method that does detection and interpolation at the same time.

        saturationDetection(exposure, saturation, doMask=True, maskName=maskName)
        if interpolate:
            saturationInterpolation(exposure, fwhm, growFootprints=growFootprints, maskName=maskName)
        if display:
            ds9.mtv(exposure.getmaskedImage(), frame=0)


    def biasCorrection(exposure, bias):

        mi = exposure.getMaskedImage()
        bmi = bias.getMaskedImage()
        mi -= bmi

    def darkCorrection(exposure, dark, expscaling, darkscaling):

        scale = expscaling / darkscaling
        mi  = exposure.getMaskedImage()
        mi.scaledMinus(scale, dark.getMaskedImage())

    def updateVariance(exposure):
        mi = exposure.getMaskedImage()
        var = afwImage.ImageF(mi.getImage(), True)
        amp = cameraGeom.cast_Amp(exposure.getDetector())
        ep = amp.getElectronicParams()
        gain = ep.getGain()
        var /= gain
        mi = afwImage.makeMaskedImage(mi.getImage(), mi.getMask(), var)
        exposure.setMaskedImage(mi)



    def flatCorrection(exposure, flat, scalingtype, scaling = 1.0):

        flatscaling = 1.0
        # Figure out scaling from the data
        # I'm not sure we should be doing this here, but maybe
        if scalingtype == 'MEAN':
            flatscaling = afwMath.makeStatistics(flat.getMaskedImage().getImage(), afwMath.MEAN).getValue(afwMath.MEAN)
        elif scalingtype == 'MEDIAN':
            flatscaling = afwMath.makeStatistics(flat.getMaskedImage().getImage(), afwMath.MEDIAN).getValue(afwMath.MEDIAN)
        elif scalingtype == 'USER':
             flatscaling = scaling
        else:
            raise pexExcept.LsstException, '%s : %s not implemented' % ("flatCorrection", scalingtype)
        mi   = exposure.getMaskedImage()
        fmi  = flat.getMaskedImage()
        
        mi.scaledDivides(1./flatscaling, fmi)

        if display:
            ds9.mtv(mi, title="Flattened")

    def illuminationCorrection(exposure, illum, illumscaling):

        # common input test

        mi   = exposure.getMaskedImage()
        mi.scaledDivides(1./illumscaling, illum.getMaskedImage())



    def trimAmp(exposure, trimBbox=None):
        """
        This returns a new Exposure that is a subsection of the input exposure.

        NOTE : do we need to deal with the WCS in any way, shape, or form?
        """
        if trimBbox is not None:
            return exposureFactory(exposure, trimBbox, LOCAL)
        else:
            amp = cameraGeom.cast_Amp(exposure.getDetector())
            return exposureFactory(exposure, amp.getDiskDataSec(false), LOCAL)
        # n.b. what other changes are needed here?
        # e.g. wcs info, overscan, etc


    def overscanCorrection(exposure, overscanBBox, fittype='MEDIAN', polyorder=1, imageFactory=afwImage.ImageF):
        """
        """

        # common input test
        mi = exposure.getMaskedImage()

        # if "True", do a deep copy
        overscanData = imageFactory(exposure.getMaskedImage().getImage(), overscanBBox, False)

        # what type of overscan modeling?
        offset = 0
        if fittype == 'MEAN':
            offset = afwMath.makeStatistics(overscanData, afwMath.MEAN).getValue(afwMath.MEAN)
            mi    -= offset
        elif fittype == 'MEDIAN':
            offset = afwMath.makeStatistics(overscanData, afwMath.MEDIAN).getValue(afwMath.MEDIAN)
            mi    -= offset
        elif fittype == 'POLY':
            raise pexExcept.LsstException, '%s : %s not implemented' % ("overscanCorrection", fittype)
        else:
            raise pexExcept.LsstException, '%s : %s an invalid overscan type' % ("overscanCorrection", fittype)

    def fringeCorrection(exposure, fringe):

        raise pexExcept.LsstException, '%s not implemented' % ("ipIsr.fringCorrection")


    def pupilCorrection(exposure, pupil):

        raise pexExcept.LsstException, '%s not implemented' % (stageName)
